### PR TITLE
refactor(http): prepare for `noUncheckedIndexedAccess`

### PR DESCRIPTION
**Pull request by gabelluardo**
Ref: #4040

### CODE REVIEW

**Comment by iuioiua** [comment]
Hi Gabriele, thank you for the PR. PTAL at https://github.com/denoland/deno_std/pull/4275#pullrequestreview-1861556153. It details how these PRs for #4040 should be approached.

**Comment by gabelluardo** [comment]
Hi @iuioiua, thanks for the suggestion, I think I did it better now.

However, I have a doubt about these two functions:

https://github.com/denoland/deno_std/blob/083a2b12ba45381c78dbabfcf06d48e2da27f16b/http/_negotiation/media_type.ts#L54-L69

and 

https://github.com/denoland/deno_std/blob/083a2b12ba45381c78dbabfcf06d48e2da27f16b/http/_negotiation/media_type.ts#L71-L86

I tested with unit tests and seems theses two function can be replaced with `accept.split(",");` and `str.split(";").map((p) => p.trim())`. Is a test bug or we can remove the `for` loops?

File: http/_negotiation/encoding.ts [hunk 1]
```diff
@@ -158,5 +158,5 @@ export function preferredEncodings(
   return priorities
     .filter(isQuality)
     .sort(compareSpecs)
-    .map((priority) => provided[priorities.indexOf(priority)]);
+    .map((priority) => provided[priorities.indexOf(priority)]!);
 }
```
Comment (iuioiua) [inline http/_negotiation/encoding.ts:162]: Can you please justify this non-null assertion?

File: http/_negotiation/language.ts [hunk 2]
```diff
@@ -140,5 +144,5 @@ export function preferredLanguages(
   return priorities
     .filter(isQuality)
     .sort(compareSpecs)
-    .map((priority) => provided[priorities.indexOf(priority)]);
+    .map((priority) => provided[priorities.indexOf(priority)]!);
```
Comment (iuioiua) [inline http/_negotiation/language.ts:147]: Ditto. Please justify this non-null assertion.

File: http/_negotiation/media_type.ts [hunk 3]
```diff
@@ -233,5 +236,5 @@ export function preferredMediaTypes(
   return priorities
     .filter(isQuality)
     .sort(compareSpecs)
-    .map((priority) => provided[priorities.indexOf(priority)]);
+    .map((priority) => provided[priorities.indexOf(priority)]!);
```
Comment (iuioiua) [inline http/_negotiation/media_type.ts:239]: Ditto.

**Comment by iuioiua** [comment]
> Hi @iuioiua, thanks for the suggestion, I think I did it better now.
> 
> However, I have a doubt about these two functions:
> 
> https://github.com/denoland/deno_std/blob/083a2b12ba45381c78dbabfcf06d48e2da27f16b/http/_negotiation/media_type.ts#L54-L69
> 
> and
> 
> https://github.com/denoland/deno_std/blob/083a2b12ba45381c78dbabfcf06d48e2da27f16b/http/_negotiation/media_type.ts#L71-L86
> 
> I tested with unit tests and seems theses two function can be replaced with `accept.split(",");` and `str.split(";").map((p) => p.trim())`. Is a test bug or we can remove the `for` loops?

Yeah, you're mostly right. Each of these can be replaced by `header.split(delimiter).map((p) => p.trim())`

File: http/server_test.ts [hunk 4]
```diff
@@ -155,7 +155,7 @@ Deno.test(
 
     try {
       assertEquals(server.addrs.length, 3);
-      assertEquals(server.addrs[0].transport, "tcp");
+      assertEquals((server.addrs[0] as Deno.NetAddr).transport, "tcp");
```
Comment (iuioiua) [inline http/server_test.ts:158]: We can use non-null assertions here.

File: http/server_test.ts [hunk 5]
```diff
@@ -164,7 +164,7 @@ Deno.test(
         (server.addrs[0] as Deno.NetAddr).port,
         listenerOneOptions.port,
       );
-      assertEquals(server.addrs[1].transport, "tcp");
+      assertEquals((server.addrs[1] as Deno.NetAddr).transport, "tcp");
```
Comment (iuioiua) [inline http/server_test.ts:167]: Ditto

File: http/server_test.ts [hunk 6]
```diff
@@ -173,7 +173,7 @@ Deno.test(
         (server.addrs[1] as Deno.NetAddr).port,
         listenerTwoOptions.port,
       );
-      assertEquals(server.addrs[2].transport, "tcp");
+      assertEquals((server.addrs[2] as Deno.NetAddr).transport, "tcp");
```
Comment (iuioiua) [inline http/server_test.ts:176]: Ditto

File: http/server_test.ts [hunk 7]
```diff
@@ -973,14 +973,14 @@ Deno.test(
     // the expected backoff delay.
     for (let i = 0; i < rejectionCount; i++) {
       assertEquals(
-        listener.acceptCallIntervals[i] >= expectedBackoffDelays[i],
+        listener.acceptCallIntervals[i]! >= expectedBackoffDelays[i]!,
```
Comment (iuioiua) [inline http/server_test.ts:976]: Can you please justify these non-null assertions?

**Review by iuioiua — CHANGES_REQUESTED** [review]

File: http/_negotiation/encoding.ts (hunk 1 above)
Comment (gabelluardo) [reply http/_negotiation/encoding.ts:162]: `priorities` is generated from the values of `provided` so `provided[priorities.indexOf(priority)]` is always non-null

File: http/_negotiation/language.ts (hunk 2 above)
Comment (gabelluardo) [reply http/_negotiation/language.ts:147]: Same here

File: http/_negotiation/media_type.ts (hunk 3 above)
Comment (gabelluardo) [reply http/_negotiation/media_type.ts:239]: And here

File: http/server_test.ts (hunk 7 above)
Comment (gabelluardo) [reply http/server_test.ts:976]: `rejectionCount` is the length of `expectedBackoffDelays` and both are not modified during the execution. The length of `listener.acceptCallIntervals` is `rejectionCount` by the initialization

File: http/server_test.ts (hunk 4 above)
Comment (gabelluardo) [reply http/server_test.ts:158]: For the other parameters is used the casting, so I have to modify them too? Just for code coherence

**Review by iuioiua — APPROVED** [review]
LGTM. @kitsonk, please feel free to review as this code was introduced to the codebase by you.

**Pull request merged**